src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,288 @@
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  assert(update_params != true);  // should never be true
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_Eta();
+    sample_mu();
+    sample_psi();
+    sample_Lambda();
+  }
+}
+
+void MFAHierarchy::sample_Eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd Sigmaeta(
+      (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+       state.Lambda.transpose() *
+           Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal()) *
+           state.Lambda)
+          .inverse());
+
+  Eigen::MatrixXd temp(state.psi.cwiseInverse().asDiagonal());
+  /*
+  std::cout << "Sigmaeta: " << Sigmaeta.rows() << ", " << Sigmaeta.cols() << std::endl;
+  std::cout << "temp: " << temp.rows() << ", " << temp.cols() << std::endl;
+  std::cout << "lambda: " << state.Lambda.rows() << ", " << state.Lambda.cols() << std::endl;
+  Eigen::VectorXd difference = data[0] - state.mu;
+  std::cout << "difference: " << difference.size() << std::endl;
+
+  std::cout << stan::math::multi_normal_rng(
+        Sigmaeta * (state.Lambda.transpose()) * 
+            temp * 
+            difference, Sigmaeta, rng) << std::endl;
+
+  std::cout << "Eta: " << state.Eta.rows() << ", " << state.Eta.cols() << std::endl;
+  */
+  // DA SISTEMARE INIZIALLIZZANDO OGNI VOLTA CORRETTAMENTE GLI ELEMENTI MANCANTI
+  if (state.Eta.rows() != card) {
+    state.Eta.resize(card, state.Eta.cols());
+    // PER ORA INIZIALIZZO COSI'
+    /*
+    for (size_t i = 0; i < card; i++) {
+      for (size_t j = 0; j < state.Eta.cols(); j++) {
+        state.Eta(i, j) = stan::math::normal_rng(0, 1, rng);
+      }
+    }
+    */
+    state.Eta = Eigen::MatrixXd::Zero(card, state.Eta.cols());
+  }
+  for (size_t i = 0; i < card; i++) {
+    state.Eta.row(i) = (stan::math::multi_normal_rng(
+        Sigmaeta * (state.Lambda.transpose()) * 
+            temp * 
+            (data[i] - state.mu),
+        Sigmaeta, rng));
+  }
+}
+
+void MFAHierarchy::sample_mu() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd Sigmamu =
+      (hypers->phi * Eigen::MatrixXd::Identity(p, p) +
+
+       card *
+           Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal())).inverse();
```

Comment:
You call `state.psi.cwiseInverse().asDiagonal()` quite often; we might store it in the `State` struct and update it when needed!